### PR TITLE
Time sensitive actions

### LINKED ISSUE #1966: Time Sensitive Actions
Time sensitive actions.

### PR DESCRIPTION
**Pull request by GrahamCampbell**
Replaces #1964. Closes #1966.

### CODE REVIEW

File: app/Foundation/Providers/IntegrationServiceProvider.php [hunk 1]
```diff
@@ -101,6 +105,20 @@ protected function registerSystem()
...
 
     /**
+     * Register the timed action instance class.
+     *
+     * @return void
+     */
+    protected function registerTimedAction()
+    {
+        $this->app->singleton(TimedActionContract::class, function (Container $app) {
+            $dates = app()->make(DateFactory::class);
```
Comment (GrahamCampbell) [inline app/Foundation/Providers/IntegrationServiceProvider.php]: Please don't use the `app` helper in service providers.

File: app/Http/Controllers/Api/ActionController.php [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+     * Get all action instances.
+     *
+     * @param \CachetHQ\Cachet\Models\TimedAction $action
+     *
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function getActionInstances(TimedAction $action)
+    {
+        // $instances = $action->instances();
+        // Graham do this
```
Comment (GrahamCampbell) [inline app/Http/Controllers/Api/ActionController.php]: :trollface:

File: app/Actions/WindowFactory.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+     * @return \Carbon\Carbon
+     */
+    protected function currentWindowStart(TimedActionModel $action)
+    {
+        $now = Carbon::now();
+
+        $diff = $action->start_at->diffInSeconds($now);
+
+        if ($diff >= 0)
+            throw new ActionNotStartedException("The timed action is only due to start in {$diff} seconds");
```
Comment (jbrooksuk) [inline app/Actions/WindowFactory.php]: Rephrase this to `The action is due to start in `{$diff} seconds`?

Comment (GrahamCampbell) [inline app/Actions/WindowFactory.php]: Hmm, I prefer the current phrasing. Makes it sound like an error instead of just a statement.

File: tests/Bus/Events/TimedAction/TimedActionWasUpdatedEventTest.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace CachetHQ\Tests\Cachet\Bus\Events\TimedAction;
+
+use CachetHQ\Cachet\Bus\Events\TimedAction\TimedActionWasUpdatedEvent;
+use CachetHQ\Cachet\Models\TimedAction;
+
+/**
+ * This is the timed action was updated event test.
+ *
+ * @author James Brooks <james@bluebaytravel.co.uk>
```
Comment (GrahamCampbell) [inline tests/Bus/Events/TimedAction/TimedActionWasUpdatedEventTest.php]: oops

Comment (jbrooksuk) [inline tests/Bus/Events/TimedAction/TimedActionWasUpdatedEventTest.php]: I've typed it all day 👊

File: app/Bus/Handlers/Commands/TimedAction/CreateTimedActionCommandHandler.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+     *
+     * @return \CachetHQ\Cachet\Models\TimedAction
+     */
+    public function handle(CreateTimedActionCommand $command)
+    {
+        $timedAction = TimedAction::create([
+            'name'                  => $command->name,
+            'description'           => $command->description,
+            'active'                => $command->active,
+            'start_at'              => $this->dates->create('d/m/Y H:i', $command->start_at),
```
Comment (jbrooksuk) [inline app/Bus/Handlers/Commands/TimedAction/CreateTimedActionCommandHandler.php]: @GrahamCampbell can you take a look at this please?

**Comment by GrahamCampbell** [comment]
Ready for final review.

**Comment by GrahamCampbell** [comment]
Ok, ready for 2nd final review. :trollface:

**Comment by jbrooksuk** [comment]
> Ok, ready for 2nd final review. :trollface:

Charts aren't quite ready yet. They're pretty hard to get right at the moment.

File: app/Presenters/TimedActionInstancePresenter.php [hunk 6]
```diff
@@ -66,9 +93,12 @@ public function did_complete_on_time()
     public function toArray()
     {
         return array_merge($this->wrappedObject->toArray(), [
-            'did_complete_on_time' => $this->did_complete_on_time(),
-            'created_at'           => $this->created_at(),
-            'updated_at'           => $this->updated_at(),
+            // 'did_complete_on_time' => $this->did_complete_on_time(),
```
Comment (GrahamCampbell) [inline app/Presenters/TimedActionInstancePresenter.php]: :trollface:

File: app/Presenters/TimedActionInstancePresenter.php [hunk 7]
```diff
@@ -23,24 +23,51 @@
  */
 class TimedActionInstancePresenter extends BasePresenter implements Arrayable
 {
+    use TimestampsTrait;
+
+    /**
+     * Present formatted date time.
```
Comment (GrahamCampbell) [inline app/Presenters/TimedActionInstancePresenter.php]: these descriptions are a bit :trollface: tbh

File: app/Http/Controllers/StatusPageController.php [hunk 8]
```diff
@@ -179,16 +179,16 @@ public function getMetrics(Metric $metric)
      */
     public function getActions(TimedAction $action)
     {
-        $actionData = [];
-
         $instances = $action->instances()->where('started_at', '>=', Carbon::now()->subDays(30))->orderBy('created_at', 'desc')->limit(30)->get();
-
```
Comment (GrahamCampbell) [inline app/Http/Controllers/StatusPageController.php]: 😲

File: app/Http/Controllers/StatusPageController.php [hunk 9]
```diff
@@ -188,7 +188,7 @@ public function getActions(TimedAction $action)
                 'started_at'           => $instance->started_at->format('Y-m-d H:i'),
                 'ended_at'             => $instance->ended_at->format('Y-m-d H:i'),
                 'completed_at'         => $instance->is_completed ? $instance->completed_at->format('Y-m-d H:i') : null,
-                'target_completed_at' => $instance->target_completed_at->format('Y-m-d H:i'),
+                'target_completed_at'  => $instance->target_completed_at->format('Y-m-d H:i'),
```
Comment (GrahamCampbell) [inline app/Http/Controllers/StatusPageController.php]: all these arrows are one space too far right

Comment (GrahamCampbell) [inline app/Http/Controllers/StatusPageController.php]: :trollface: :shipit:

Comment (jbrooksuk) [inline app/Http/Controllers/StatusPageController.php]: 🚢

**Comment by GrahamCampbell** [comment]
TODO:

* [ ] Add validation on the completed_at time.
* [ ] Fix the tests
* [ ] Deal with late points
* [ ] Deal with not completed

**Comment by ConnorVG** [comment]
Ready for third final review? :trollface:

File: app/Http/Controllers/Api/ActionController.php [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function getActions()
+    {
+        $actions = TimedAction::active();
+
+        $actions->search(Binput::except(['sort', 'order', 'per_page']));
+
+        if ($createdAfter = Binput::get('created_after')) {
+            $actions->whereDate('created_at', '>', Carbon::createFromFormat('Y-m-d', $createdAfter)->toDateString());
```
Comment (GrahamCampbell) [inline app/Http/Controllers/Api/ActionController.php]: We might want to use equal to or greater than here.

Comment (jbrooksuk) [inline app/Http/Controllers/Api/ActionController.php]: 👍 I did think about that.

File: app/Http/Controllers/Api/ActionController.php [hunk 11]
```diff
@@ -43,9 +43,9 @@ public function getActions()
         $actions->search(Binput::except(['sort', 'order', 'per_page']));
 
         if ($createdAfter = Binput::get('created_after')) {
-            $actions->whereDate('created_at', '>', Carbon::createFromFormat('Y-m-d', $createdAfter)->toDateString());
+            $actions->whereDate('created_at', '>=', Carbon::createFromFormat('Y-m-d', $createdAfter)->toDateString());
         } elseif ($createdBefore = Binput::get('created_before')) {
-            $actions->whereDate('created_at', '<', Carbon::createFromFormat('Y-m-d', $createdAfter)->toDateString());
+            $actions->whereDate('created_at', '<=', Carbon::createFromFormat('Y-m-d', $createdAfter)->toDateString());
```
Comment (ConnorVG) [inline app/Http/Controllers/Api/ActionController.php]: This is just confusing. If you put exactly the same date for 'after' or 'before' and something was made on _that date_, you end up with that same item in both ranges. Not really... Expected?

idk, seems weird. Maybe just a wording issue, ie 'on or after' + 'before'.

File: app/Bus/Commands/TimedAction/CreateTimedActionCommand.php [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+     * @param bool     $visible
+     * @param string   $start_at
+     * @param string   $timezone
+     * @param int      $schedule_interval
+     * @param int      $completion_latency
+     * @param int|null $timed_action_group_id
+     *
+     * @return void
+     */
+    public function __construct($name, $description, $active, $visible, $start_at, $timezone, $schedule_interval, $completion_latency, $timed_action_group_id = null)
```
Comment (ConnorVG) [inline app/Bus/Commands/TimedAction/CreateTimedActionCommand.php]: :scream:

Comment (jbrooksuk) [inline app/Bus/Commands/TimedAction/CreateTimedActionCommand.php]: ?

File: resources/views/partials/actions.blade.php [hunk 13]
```diff
@@ -88,7 +88,21 @@ function drawChart($el) {
                             return "{{ color_darken($theme_reds, -0.1) }}";
                         }),
                         pointBorderColor: "{{ color_darken($theme_metrics, -0.1) }}",
-                        pointHoverBackgroundColor: "{{ color_darken($theme_metrics, -0.2) }}",
```
Comment (ConnorVG) [inline resources/views/partials/actions.blade.php]: I feel like this logic should be moved to the JS import and provide a function for registering. 

JS in the wild is ugly.

Comment (jbrooksuk) [inline resources/views/partials/actions.blade.php]: Metrics are the same. It'll be moved around with the Vue stuff anyway.

Comment (ConnorVG) [inline resources/views/partials/actions.blade.php]: :scream: :sob:

File: resources/views/partials/actions.blade.php [hunk 14]
```diff
@@ -89,6 +89,7 @@ function drawChart($el) {
                                 stepSize: 3600,
                                 min: 0,
                                 callback: function (value, index, values) {
+                                    console.log(value, index, values)
```
Comment (ConnorVG) [inline resources/views/partials/actions.blade.php]: what the fuck

Comment (jbrooksuk) [inline resources/views/partials/actions.blade.php]: It's for debugging 😢

Comment (ConnorVG) [inline resources/views/partials/actions.blade.php]: Then why is it in Git? `:facepalm:`

Comment (jbrooksuk) [inline resources/views/partials/actions.blade.php]: We'll be squashing all the commits in the future anyway.

Comment (ConnorVG) [inline resources/views/partials/actions.blade.php]: :trollface:

File: app/Bus/Commands/TimedAction/CreateTimedActionCommand.php (hunk 12 above)
Comment (yoyosan) [inline app/Bus/Commands/TimedAction/CreateTimedActionCommand.php]: That number of parameters though :trollface:

Comment (ConnorVG) [inline app/Bus/Commands/TimedAction/CreateTimedActionCommand.php]: Right!
